openssl/src/argon2.rs

```diff
@@ -0,0 +1,147 @@
+use libc::c_void;
+use std::ffi::CStr;
+use std::ptr;
+
+use crate::error::ErrorStack;
+use crate::{cvt, cvt_p};
+
+/// Derives a key using the argon2id algorithm.
+///
+/// Requires OpenSSL 3.2.0 or newer.
+#[allow(clippy::too_many_arguments)]
+pub fn argon2id(
+    pass: &[u8],
+    salt: &[u8],
+    ad: Option<&[u8]>,
+    secret: Option<&[u8]>,
+    mut iter: u32,
+    mut threads: u32,
+    mut lanes: u32,
+    mut memcost: u32,
+    out: &mut [u8],
+) -> Result<(), ErrorStack> {
+    // We only support single-threaded operation for now since rust-openssl doesn't
+    // bind OSSL_set_max_threads
+    assert!(threads == 1);
+    let pass_field = CStr::from_bytes_with_nul(b"pass\0").unwrap();
+    let salt_field = CStr::from_bytes_with_nul(b"salt\0").unwrap();
+    let ad_field = CStr::from_bytes_with_nul(b"ad\0").unwrap();
+    let secret_field = CStr::from_bytes_with_nul(b"secret\0").unwrap();
+    let iter_field = CStr::from_bytes_with_nul(b"iter\0").unwrap();
+    let size_field = CStr::from_bytes_with_nul(b"size\0").unwrap();
+    let threads_field = CStr::from_bytes_with_nul(b"threads\0").unwrap();
+    let lanes_field = CStr::from_bytes_with_nul(b"lanes\0").unwrap();
+    let memcost_field = CStr::from_bytes_with_nul(b"memcost\0").unwrap();
```

Comment:
I don't think there's a huge advantage to creating all these `CStr`, you can just do `b"iter\0".as_ptr()` below